engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/ConfluentAvroToJsonFormatterFactory.scala

```diff
@@ -0,0 +1,120 @@
+  * @tparam V - value type passed from KafkaAvroSourceFactory, used to determine which datumReaderWriter use (e.g. specific or generic)
+  */
+class ConfluentAvroToJsonFormatter[K: ClassTag, V: ClassTag](kafkaConfig: KafkaConfig,
+                                                             schemaRegistryClient: SchemaRegistryClient,
+                                                             messageFormatter: ConfluentAvroMessageFormatter,
+                                                             messageReader: ConfluentAvroMessageReader,
+                                                             deserializationSchema: KafkaDeserializationSchema[ConsumerRecord[K, V]]
+                                                            ) extends RecordFormatter {
+
+  /**
+    * Step 1: Deserialize raw kafka event to GenericRecord/SpecificRecord domain.
+    * Step 2: Create Encoders that use ConfluentAvroMessageFormatter to convert avro object to json
+    * Step 3: Encode event's data with schema id's with derived encoder.
+    */
+  override protected def formatRecord(record: ConsumerRecord[Array[Byte], Array[Byte]]): Array[Byte] = {
+    val deserializedRecord = deserializationSchema.deserialize(record)
+    val serializableRecord = AvroSerializableConsumerRecord(
+      if (kafkaConfig.useStringForKey) None else Option(record.key()).map(ConfluentUtils.readId),
+      Option(record.value()).map(ConfluentUtils.readId).getOrElse(throw new IllegalArgumentException("Failed to read value schema id")),
+      SerializableConsumerRecord(deserializedRecord)
+    )
+
+    consumerRecordEncoder(serializableRecord).noSpaces.getBytes(StandardCharsets.UTF_8)
+  }
+
+  /**
+    * Step 1: Deserialize raw json bytes to AvroSerializableConsumerRecord[Json, Json] domain without interpreting key and value content.
+    * Step 2: Create key and value json-to-avro interpreter based on schema id's provided in json.
+    * Step 3: Use interpreter to create raw kafka ConsumerRecord
+    */
+  override protected def parseRecord(topic: String, bytes: Array[Byte]): ConsumerRecord[Array[Byte], Array[Byte]] = {
+    val record = CirceUtil.decodeJsonUnsafe(bytes)(consumerRecordDecoder)
+
+    def serializeKeyValue(keyOpt: Option[Json], value: Json): (Array[Byte], Array[Byte]) = {
+      val keyBytes = if (kafkaConfig.useStringForKey) {
+        keyOpt match {
+          // we handle strings this way because we want to keep result value compact and JString is formatted in quotes
+          case Some(j) if j.isString => j.asString.get.getBytes(StandardCharsets.UTF_8)
+          case None => null
+        }
+      } else {
+        val keySchema = record.keySchemaId.map(id => getSchemaById(id)).getOrElse(throw new IllegalArgumentException("Error reading schema: empty schema id"))
+        keyOpt.map(keyJson => messageReader.readJson[K](keyJson, keySchema, ConfluentUtils.keySubject(topic))).getOrElse(Array.emptyByteArray)
```

Comment:
Instead of returning empty array, please throw exception with message showing that for avro keys we expected defined, properly formatted key